audisp/plugins/af_unix/audisp-af_unix.8
.TH AUDISP-AF_UNIX "8" "Aug 2025" "Red Hat" "System Administration Utilities"
.SH NAME
audisp-af_unix \- plugin to push audit events to an af_unix socket
.SH SYNOPSIS
.B audisp-af_unix
[ \fIOPTIONS\fP ]
.SH DESCRIPTION
\fBaudisp-af_unix\fP is a plugin for the audit event dispatcher that sends audit events to an af_unix socket where other applications can read events. The
.B args
line of the 
.B af_unix.conf
file expects three arguments: access mode, socket path, and output format, and optionally a fourth argument specifying the queue depth. The access mode determines the permissions for the socket and defaults to 0640. The socket path specifies where the socket will be created, with the default location being /run/audit/audispd_events. The output format determines the format in which events are delivered to the socket and supports two options: "string" and "binary". The "string" format delivers events in a human-readable form, while the "binary" format delivers events in their binary representation, which is essential for applications that need to process events in binary and reconstruct headers accurately. If the output format is not specified, the plugin defaults to the "string" format. If no queue depth is specified, it defaults to 512.

The
.B af_unix.conf
file must also include the line \fBformat = binary\fP. This setting specifies the input format that the \fBaudisp-af_unix\fP plugin expects from the audit event dispatcher. It ensures that the input delivered to the plugin is in binary format, enabling the plugin to reconstruct headers in their proper binary structure.

When the plugin is sent \fBSIGUSR1\fP, it writes a state report to \fBaf_unix.state\fP.

.SH FILES
/etc/audit/plugins/af_unix.conf
/run/audit/audispd_events
/run/audit/af_unix.state
/etc/audit/auditd.conf
.SH "SEE ALSO"
.BR auditd.conf (8),
.BR auditd-plugins (5).
.SH AUTHOR
Steve Grubb


ChangeLog
4.1.2
- Use runstatedir to guide the whole audit project to the run directory
- Deprecate the notion that plugins can be anything other than outbound
- Bring set_aumessage_mode back into libaudit's ABI but not API
- Add anomaly, response, and crypto reports to aureport
- Drop IPX header handling and require kernel 5.15 or later
- Resolved a number of FIXME's all over the code base
- Optimize ausearch/report log processing
- In auparse normalizer, save pids as object attributes if its a process group
- Fix inbound protocol detection for af_unix plugin
- Add an internal queue to the af_unix plugin
- Both af_unix and remote plugin write status report to /run/audit/

4.1.1
- Add libauplugin example program and improve its documentation
- auditctl -A option was deleting - now prepends like it used to
- Adjust af_unix and dispatcher to send exact record size in binary mode
- Add bash completions for ausearch, aureport, auditctl, and augenrules
- ausearch/aureport: allow symlinked config
- Add support for file_getattr and file_setattr syscalls

4.1.0
- Fix auditd -s enable hang issue (Yan Zhu)
- Optimize event formatting in auditd
- af_unix plugin: Restore terminating newlines
- Add support for "exec" action in max_log_file_action in auditd
- Refactor auparse code to be multi-thread safe
- Add memory pool to netlink event processing to reduce memory churn
- Make all plugins ignore SIGTERM if not from auditd (issue #469)
- Add libauplugin, refactor audisp-filter, ids, and audisp-statsd to use it
- In auditd, safely reconfigure the network settings after SIGHUP
- Make test suite machine independent
- Persistent queue support with metrics helpers

4.0.5
- Rework audisp queue to be lockless
- Fix missing delete command in auditctl
- Allow plus addresses (rfc5233) to auditd email.
- Reduce memory churn in auditd event dispatching
- Add configurable recurring state report in auditd
- Switch audisp-statsd to stop sending signals
- Add glibc memory stats to audisp-statsd

4.0.4
- auditctl: update io_uring operations table
- update syscall table for 6.15
- auditd.cron.5: Describe time-based log rotation setup
- auditd: Broadcast a warning on startup if a system halt is possible (#435)
- Fix audisp-remote segfault on connection error (#446)
- Improve locating last event if ausearch is using checkpointing
- af_unix plugin: fix string mode support
- Remove const from audit_rule_fieldpair_data & audit_rule_interfield_comp_data
- Add various updates to the experimental ids plugin
- Add glibc memory statistics to auditd state report

4.0.3
- Remove a RHEL4 flag table since it's been unsupported for a while
- Change dependency from Requires to Wants for audit-rules.service
- Disable ProtectKernelModules by default in auditd.service
- Skip plugin configs that do not have .conf suffix
- audisp-filter: iterate records correctly when forwarding
- Update syscall table for missing syscalls
- Modify ausearch checkpoint code to address 64 inode and device numbers
- Fix potential segfault interpreting relative paths
- Add audit_set_enabled & audit_is_enabled back to the libaudit python bindings
- Log runlevel changes to console during boot (Attila Lakatos)
- Add audit-tmpfiles.conf to ensure /var/log/audit exists (Colin Walters)
- Propagate event format to the audisp-af_unix plugin (Attila Lakatos)
- Add support for RISC-V - riscv32, riscv64 (David Abdurachmanov)

4.0.2
- Fix musl C builds
- Many code cleanups (Yugend)
- Use atomic variables if available for signal related flags
- Dont rotate audit logs when auditd is in debug mode
- Fix a couple memory leaks on error paths
- Correct output when displaying rules with exe/path/dir (Attila Lakatos)
- Fix auparse lookup test to not use the system libaupaurse
- Improve auparse metrics
- Update auparse normalizer for recent syscalls
- Make status report uniform

4.0.1
- Update TRUSTED_APP interpretation to look for known fields
- In auditd plugins, allow variable amount of arguments (Attila Lakatos)
- Fix augenrules to work correctly when kernel is in immutable mode
- Add ausearch_cur_event to auparse library (Attila Lakatos)
- Add audisp-filter plugin (Attila Lakatos)
- Improve sorting speed of aureport --summary reports
- auditd & audit-rules.service pick up paths automatically (Laurent Bigonville)
- Update auparse normalizer for new syscalls

4.0
- Drop python2 support
- Drop auvirt and autrace programs
- Drop SysVinit support
- Require the use of the 5.0 or later kernel headers
- New README.md file
- Rewrite legacy service functions in terms of systemctl
- Consolidate and update end of event detection to a common function
- Split off rule loading from auditd.service into audit-rules.service
- Refactor libaudit.h to split out logging functions and record numbers
- Speed up aureport --summary reports
- Limit libaudit python bindings to logging functions
- Add a metrics function for auparse
- Change auditctl to use pidfd_send_signal for signaling auditd
- Adjust watches to optimize syscalls hooked when watch file access
- Drop nispom rules
- Add intepretations for fsconfig, fsopen, fsmount, & move_mount
- Many code fixups (cgzones)
- Update syscall and interpretation tables to the 6.8 kernel

3.1.2
- When processing a run level change, make auditd exit
- In auditd, fix return code when rules added in immutable mode
- In auparse, when files are given, also consider EUID for access
- Auparse now interprets unnamed/anonymous sockets (Enzo Matsumiya)
- Disable Python bindings from setting rules due to swig bug (S. Trofimovich)
- Update all lookup tables for the 6.5 kernel
- Don't be as paranoid about auditctl -R file permissions
- In ausearch, correct subject/object search to be an and if both are given
- Adjust formats for 64 bit time_t
- Fix segfault in python bindings around the feed API
- Add feed_has_data, get_record_num, and get/goto_field_num to python bindings

3.1.1
- Add user friendly keywords for signals to auditctl

audisp/plugins/af_unix/audisp-af_unix.c
/*
 * af_unix.c - implementation of the audisp-af_unix plugin
 * Copyright (c) 2023 Red Hat Inc.
 * All Rights Reserved.
 *
 * This software may be freely redistributed and/or modified under the
 * terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor
 * Boston, MA 02110-1335, USA.
 *
 * Authors:
 *   Steve Grubb <sgrubb@redhat.com>
 */

#include "config.h"
#include <stdio.h>
#include <syslog.h>
#include <ctype.h>
#include <errno.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <time.h>
#ifdef HAVE_MALLINFO2
#include <malloc.h>
#endif
#include <libgen.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/uio.h>
#include <dirent.h>
#include <sys/un.h>
#include <fcntl.h>
#include <poll.h>
#include <sys/socket.h>
#ifdef HAVE_LIBCAP_NG
#include <cap-ng.h>
#endif
#include "libaudit.h"
#include "auplugin.h"
#include "audispd-pconfig.h"
#include "queue.h"

#define DEFAULT_PATH AUDIT_RUN_DIR"/audispd_events"
#define STATE_FILE AUDIT_RUN_DIR"/af_unix.state"
//#define DEBUG

/* Global Data */
static volatile int stop = 0, hup = 0, usr1 = 0;
char rx_buf[MAX_AUDIT_EVENT_FRAME_SIZE+1];
int sock = -1, conn = -1, client = 0;
struct pollfd pfd[3];
unsigned mode = 0;
format_t format = -1;
char *path = NULL;
int inbound_protocol = -1;

#ifdef HAVE_MALLINFO2
static struct mallinfo2 last_mi;
#endif

#define DEFAULT_QUEUE_DEPTH 512
#define QUEUE_ENTRY_SIZE MAX_AUDIT_EVENT_FRAME_SIZE+1

static size_t queue_depth = DEFAULT_QUEUE_DEPTH;
static struct queue *queue;
static const unsigned char *out_buf;
static size_t out_len;
static size_t out_off;

/*
 * SIGTERM handler
 *
 * Only honor the signal if it comes from the parent process so that other
 * tasks (cough, systemctl, cough) can't make the plugin exit without
 * the dispatcher in agreement. Otherwise it will restart the plugin.
 */
static void term_handler(int sig, siginfo_t *info, void *ucontext)
{
	if (info && info->si_pid != getppid())
		return;
	stop = 1;
}

/*
 * SIGHUP handler: re-read config
 */
static void hup_handler(int sig)
{
	hup = 1;
}

static void usr1_handler(int sig)
{
	usr1 = 1;
}

#ifdef HAVE_MALLINFO2
static void write_memory_state(FILE *f)
{
	struct mallinfo2 mi = mallinfo2();

	fprintf(f, "glibc arena (total memory) is: %zu KiB, was: %zu KiB\n",
		(size_t)mi.arena/1024, (size_t)last_mi.arena/1024);
	fprintf(f, "glibc uordblks (in use memory) is: %zu KiB, was: %zu KiB\n",
		(size_t)mi.uordblks/1024,(size_t)last_mi.uordblks/1024);
	fprintf(f,"glibc fordblks (total free space) is: %zu KiB, was: %zu KiB\n",
		(size_t)mi.fordblks/1024,(size_t)last_mi.fordblks/1024);

	memcpy(&last_mi, &mi, sizeof(struct mallinfo2));
}
#endif

static void write_state_report(void)
{
	char buf[64];
	mode_t u = umask(0137); // allow 0640
	FILE *f = fopen(STATE_FILE, "w");
	umask(u);
	if (f == NULL)
		return;

	time_t now = time(NULL);
	strftime(buf, sizeof(buf), "%x %X", localtime(&now));
	fprintf(f, "current time = %s\n", buf);
	fprintf(f, "client_connected = %s\n", client ? "yes" : "no");
	fprintf(f, "queue_length = %zu\n", q_queue_length(queue));
	fprintf(f, "max_queue_length = %zu\n", q_max_queue_length(queue));
	fprintf(f, "queue_size = %zu\n", q_queue_size(queue));
#ifdef HAVE_MALLINFO2
	write_memory_state(f);
#endif
	fclose(f);
}


int create_af_unix_socket(const char *spath, int mode)
{
	struct sockaddr_un addr;
	socklen_t len;
	int rc, cmd, one = 1;

	sock = socket(PF_UNIX, SOCK_STREAM, 0);
	if (sock < 0) {
		syslog(LOG_ERR, "Couldn't open af_unix socket (%s)",
		       strerror(errno));
		return -1;
	}
	setsockopt(sock, SOL_SOCKET, SO_REUSEADDR,
		   (char *)&one, sizeof (int));
#ifdef DEBUG
	printf("%o %s\n", mode, spath);
#else
	memset(&addr, 0, sizeof(addr));
	addr.sun_family = AF_UNIX;
	snprintf(&addr.sun_path[0], 108, "%.107s", spath);
	len = sizeof(addr);
	rc = bind(sock, (const struct sockaddr *)&addr, len);
	if (rc < 0) {
		syslog(LOG_ERR, "Couldn't bind af_unix socket (%s)",
		       strerror(errno));
		close(sock);
		return -1;
	}
	rc = chmod(spath, mode);
	if (rc < 0) {
		syslog(LOG_ERR, "Couldn't chmod %s to %04o (%s)",
		       spath, mode, strerror(errno));
		close(sock);
		unlink(spath);
		return -1;
	}
	// Put socket in nonblock mode and don't leak the descriptor
	cmd = fcntl(sock, F_GETFL);
	fcntl(sock, F_SETFL, cmd|FNDELAY|FD_CLOEXEC);

	// Make socket listening...won't block
	(void)listen(sock, 1);
#endif
	return 0;
}

int setup_socket(int argc, char *argv[])
{
	for (int i = 1; i < argc; i++) {
		char *arg = argv[i];
		if (isdigit((unsigned char)arg[0])) {
			errno = 0;
			if (mode == 0) {
				// parse mode
				mode = strtoul(arg, NULL, 8);
				if (errno) {
					syslog(LOG_ERR,
						"Error converting %s (%s)",
						argv[i], strerror(errno));
					mode = 0;
				}
			} else {
				// parse queue depth
				queue_depth = strtoul(arg, NULL, 10);
				if (errno || queue_depth == 0) {
					syslog(LOG_ERR,
						"Error converting %s (%s)",
						argv[i], strerror(errno));
					queue_depth = DEFAULT_QUEUE_DEPTH;
				}
			}
		} else if (strchr(arg, '/') != NULL) {
			// parse path
			char* base;
			path = arg;
			// Make sure there are directories
			base = strchr(path, '/');
			if (base) {
				DIR* d;
				char* dir = strdup(path);
				base = dirname(dir);
				d = opendir(base);
				if (d) {
					closedir(d);
					free(dir);
				} else {
					syslog(LOG_ERR,
					       "Couldn't open %s (%s)",
					       base, strerror(errno));
					free(dir);
					exit(1);
				}
			} else {
				syslog(LOG_ERR, "Malformed path %s",
				       path);
				exit(1);
			}
		} else {
			if (strcmp(arg, "string") == 0)
				format = F_STRING;
			else if (strcmp(arg, "binary") == 0)
				format = F_BINARY;
			else
				syslog(LOG_ERR, "Invalid format detected");
		}
	}

	if (mode == 0 || path == NULL || format == -1) {
		syslog(LOG_ERR, "Bad or not enough arguments, using defaults");
		if (mode == 0) {
			mode = 0640;
			syslog(LOG_INFO, "Using default mode");
		}
		if (path == NULL) {
			path = DEFAULT_PATH;
			syslog(LOG_INFO, "Using default path");
		}
		if (format == -1) {
			format = F_STRING;
			syslog(LOG_INFO, "Using default format");
		}
	}
	if (queue_depth == DEFAULT_QUEUE_DEPTH)
		syslog(LOG_INFO, "Using default queue depth");

	return create_af_unix_socket(path, mode);
}

static int event_to_string(struct audit_dispatcher_header *hdr,
			   char *data, char **out, int *outlen)
{
	char *v = NULL, *ptr, unknown[32];
	int len;

	if (inbound_protocol == F_BINARY) {
		if (hdr->ver == AUDISP_PROTOCOL_VER) {
			const char *type;

			/* Get the event formatted */
			type = audit_msg_type_to_name(hdr->type);
			if (type == NULL) {
				snprintf(unknown, sizeof(unknown),
					 "UNKNOWN[%u]", hdr->type);
				type = unknown;
			}
			len = asprintf(&v, "type=%s msg=%.*s\n",
				       type, hdr->size, data);
		} else if (inbound_protocol == F_BINARY &&
			   hdr->ver == AUDISP_PROTOCOL_VER2) {
			// Protocol 2 events are already formatted
			len = asprintf(&v, "%.*s\n", hdr->size, data);
		} else
			len = 0;
	} else if (inbound_protocol == F_STRING) {
		// Inbound strings start at the hdr
		len = asprintf(&v, "%s\n", (char *)hdr);
	} else
		len = 0;
	if (len <= 0) {
		*out = NULL;
		*outlen = 0;
		return -1;
	}

	/* Strip newlines from event record except the last one */
	ptr = v;
	while ((ptr = strchr(ptr, 0x0A)) != NULL) {
		if (ptr != &v[len-1])
			*ptr = ' ';
		else
			break; /* Done - exit loop */
	}

	*out = v;
	*outlen = len;
	return 1;
}

/*
 * read_binary_record - read a binary dispatcher record
 * @fd:  input descriptor
 * @hdr: pointer to header storage
 * @data: pointer to data storage
 *
 * This function reads exactly sizeof(*hdr) bytes followed by hdr->size
 * bytes from @fd.  It returns the total bytes read or -1 on error and
 * 0 when EOF is reached.
 */
static int read_binary_record(int fd, struct audit_dispatcher_header *hdr,
			      char *data)
{
	size_t len = sizeof(*hdr);
	char *ptr = (char *)hdr;
	ssize_t rc;

	while (len) {
		rc = read(fd, ptr, len);
		if (rc <= 0) {
			if (rc < 0 && errno == EINTR)
				continue;
			return rc;
		}
		ptr += rc;
		len -= rc;
	}

	if (hdr->size > MAX_AUDIT_MESSAGE_LENGTH)
		hdr->size = MAX_AUDIT_MESSAGE_LENGTH;

	len = hdr->size;
	ptr = data;
	while (len) {
		rc = read(fd, ptr, len);
		if (rc <= 0) {
			if (rc < 0 && errno == EINTR)
				continue;
			return rc;
		}
		ptr += rc;
		len -= rc;
	}

	return sizeof(*hdr) + hdr->size;
}

int failed_append = 0;
void read_audit_record(int ifd)
{
	int len;

	// If it's the first call, detect which inbound protocol we are using
	if (inbound_protocol == -1) {
		unsigned char peek[4];
		ssize_t rc;

		// audisp uses socketpair to setup stdin, use recvfrom to
		// peek into what the protocol might be
		rc = recvfrom(ifd, peek, sizeof(peek), MSG_PEEK, NULL, NULL);
		if (rc < 0) {
			if (errno == ENOTSOCK) {
				// Let's assume it is a file which means string
				inbound_protocol = F_STRING;
				goto resolved;
			}
			syslog(LOG_ERR, "recvfrom failed (%s)",
			       strerror(errno));
			stop = 1;
			return;
		}
		if (rc == 0) {
			stop = 1;
			return;
		}
		// If its string, there will be an ascii letter in the first
		// character. If it's binary, it will be a 1 (0 is ancient)
		if (peek[0] == 1 || peek[0] == 0)
			inbound_protocol = F_BINARY;
		else
			inbound_protocol = F_STRING;
	}

resolved:
	if (inbound_protocol == F_BINARY) {
		struct audit_dispatcher_header *hdr =
			(struct audit_dispatcher_header *)rx_buf;
		char *data = rx_buf + sizeof(*hdr);

		len = read_binary_record(ifd, hdr, data);
		if (len <= 0) {
			if (len == 0)
				stop = 1;
			return;
		}

		if (client && !stop) {
			if (format == F_STRING) {
				char *str = NULL;
				int str_len = 0;

				if (event_to_string(hdr, data, &str,
						    &str_len) < 0)
					return;

				if (q_append(queue, str, str_len) != 0) {
					if (failed_append < 5)
						syslog(LOG_ERR,
				  "Dropping event - queue append failed (%s)",
						       strerror(errno));
					failed_append++;
				} else
					failed_append = 0;
				free(str);
			} else if (format == F_BINARY) {
				int total = sizeof(*hdr) + hdr->size;
				char *buf = malloc(total);
				if (buf) {
					memcpy(buf, hdr, sizeof(*hdr));
					memcpy(buf + sizeof(*hdr), data,
					       hdr->size);
					if (q_append(queue, buf, total) != 0) {
						if (failed_append < 5)
							syslog(LOG_ERR,
				  "Dropping event - queue append failed (%s)",
							   strerror(errno));
						failed_append++;
					} else
						failed_append = 0;
					free(buf);
				}
			}
		}
	} else {
		do {
			len = auplugin_fgets(rx_buf,
					   MAX_AUDIT_EVENT_FRAME_SIZE + 1, ifd);
			if (len > 0) {
				if (client && !stop) {
					if (format == F_STRING) {
						if (q_append(queue, rx_buf,
							     len) != 0)
							syslog(LOG_ERR,
						    "Queue append failed (%s)",
							       strerror(errno));
					} else if (format == F_BINARY) {
						struct audit_dispatcher_header hdr;

						hdr.ver = AUDISP_PROTOCOL_VER2;
						hdr.hlen = sizeof(struct audit_dispatcher_header);
						hdr.type = 0;
						hdr.size = len;
						int total = sizeof(hdr) + len;
						char *buf = malloc(total);
						if (buf) {
							memcpy(buf, &hdr,
							       sizeof(hdr));
							memcpy(buf+sizeof(hdr),
							       rx_buf, len);
							if (q_append(queue, buf,
								    total) != 0)
								syslog(LOG_ERR,
						     "Queue append failed (%s)",
							       strerror(errno));
							free(buf);
						}
					}
				}
			} else if (auplugin_fgets_eof())
				stop = 1;
		} while (!stop &&
			 auplugin_fgets_more(MAX_AUDIT_EVENT_FRAME_SIZE));
	}
}

void accept_connection(void)
{
	int tmp_conn;

	do {
		tmp_conn = accept4(sock, NULL,NULL, SOCK_NONBLOCK|SOCK_CLOEXEC);
	} while (tmp_conn < 0 && errno == EINTR);

	if (tmp_conn >= 0) {
		if (conn < 0) {
			syslog(LOG_INFO, "Client connected");
			client = 1;
			conn = tmp_conn;
		} else
			close(tmp_conn);
	}
}

static void send_queue(void)
{
	ssize_t rc;

	while (!q_empty(queue) && client && !stop) {
		if (out_off == 0) {
			int r = q_peek(queue, &out_buf, &out_len);
			if (r <= 0) {
				if (r < 0)
					syslog(LOG_ERR,
					       "Queue peek failed (%s)",
					       strerror(errno));
				return;
			}
		}

		do {
			rc = write(conn, out_buf + out_off,
				   out_len - out_off);
		} while (rc < 0 && errno == EINTR);
		if (rc < 0) {
			if (errno == EAGAIN)
				return;
			if (errno == EPIPE) {
				close(conn);
				conn = -1;
				client = 0;
				auplugin_fgets_clear();
				out_off = 0;
				out_len = 0;
				out_buf = NULL;
			}
			return;
		}
		out_off += rc;
		if (out_off == out_len) {
			q_drop_head(queue);
			out_off = 0;
			out_len = 0;
			out_buf = NULL;
		}
	}
}

void event_loop(int ifd)
{
	// setup poll
	pfd[0].fd = ifd;	//stdin
	pfd[0].events = POLLIN;
	pfd[1].fd = sock;	// listen socket
	pfd[1].events = POLLIN;

	// loop on poll until stop - not doing HUP for now
	while (!stop) {
		int rc;

		if (usr1) {
			write_state_report();
			usr1 = 0;
		}

		if (client) {
			pfd[2].fd = conn;       // the client
			pfd[2].events = POLLHUP;
			if (!q_empty(queue))
				pfd[2].events |= POLLOUT;
		}

		rc = poll(pfd, 2 + client, -1);
		if (rc < 0) {
			if (stop)
				break;

			if (errno == EINTR)
				continue;

			syslog(LOG_WARNING, "Poll error (%s), exiting",
			       strerror(errno));
			return;
		}
		if (rc > 0) {
			if (client && (pfd[2].revents & POLLHUP)) {
				// client hung up, do this first in case
				// an inbound audit record is available
				close(conn);
				conn = -1;
				client = 0;
				auplugin_fgets_clear();
				out_off = 0;
				out_len = 0;
				out_buf = NULL;
			}
			// auditd closed it's socket, exit
			if (pfd[0].revents & POLLHUP) {
				syslog(LOG_INFO,
				       "Auditd closed it's socket - exiting");
				return;
			}
			if (pfd[0].revents & POLLIN) {
				// Inbound audit event
				read_audit_record(ifd);
			}

			if (pfd[1].revents & POLLIN) {
				// someone connected, accept it
				accept_connection();
				send_queue();
			}

			if (client && (pfd[2].revents & POLLOUT))
				send_queue();
		}
	}
	if (stop == 1)
		syslog(LOG_INFO,
		       "audisp-af_unix plugin is exiting on TERM request");
}


int main(int argc, char *argv[])
{
	struct sigaction sa;
	int i, ifd;

	/* Register sighandlers */
	sa.sa_flags = 0;
	sigemptyset(&sa.sa_mask);
	/* Ignore all signals by default */
	sa.sa_handler = SIG_IGN;
	for (i=1; i<NSIG; i++)
		sigaction( i, &sa, NULL );

	/* Set handler for the ones we care about */
	sa.sa_handler = hup_handler;
	sa.sa_flags = 0;
	sigaction(SIGHUP, &sa, NULL);
	sa.sa_handler = usr1_handler;
	sigaction(SIGUSR1, &sa, NULL);
	sa.sa_sigaction = term_handler;
	sa.sa_flags = SA_SIGINFO;
	sigaction(SIGTERM, &sa, NULL);
	/* Set STDIN non-blocking */
	(void) umask( umask( 077 ) | 027 );
	ifd = 0;
#ifdef DEBUG
	ifd = open("test.log", O_RDONLY);
#endif
	fcntl(ifd, F_SETFL, O_NONBLOCK);

	// Initialize the socket
	if (setup_socket(argc, argv)) {
		syslog(LOG_ERR,"audisp-af_unix plugin exiting due to errors "
		       "setting up socket");
		exit(1);
	}

#ifdef HAVE_LIBCAP_NG
	// Drop capabilities
	capng_clear(CAPNG_SELECT_BOTH);
	if (capng_apply(CAPNG_SELECT_BOTH))
		syslog(LOG_WARNING, "audisp-af_unix plugin was unable to "
		       "drop capabilities, continuing with elevated priviles");
#endif
	queue = q_open(queue_depth, QUEUE_ENTRY_SIZE);
	if (queue == NULL) {
		syslog(LOG_ERR, "Unable to create queue (%s)",
		       strerror(errno));
		exit(1);
	}
	syslog(LOG_INFO, "audisp-af_unix plugin is listening for events");
	event_loop(ifd);

	// close up and delete socket
	if (conn >= 0) close(conn);
	if (sock >= 0) close(sock);
	if (unlink(path) == -1) {
		syslog(LOG_WARNING, "Failed to unlink socket %s (%s)",
		       path, strerror(errno));
	}
	q_close(queue);

	return 0;
}

